### ISSUE TITLE
first load of inflection tables is causing slowdown

### ISSUE THREAD

**Issue by balmas** [alpheios-project/inflection-tables#170]
the first load of the inflection tables is causing a significant slowdown for the extension. Firefox warns the user about it and offers the option to stop the script. Chrome doesn't so far, but it results in the first lookup being very slow to finish.

**Comment by kirlat** [comment]
Provided some speed optimization, mostly the ones that are more obvious and do not require extensive code changes. By my estimates, it decreased rendering time by 10 to 40 percent. Please let me know if it make that FF message go away.

If not, I will try Vue.js lazy rendering. Right now all tables are rendered at once. We might try to render each particular table at the moment users clicks on its header to open. This should not require extensive code changes too (which I would like to spare for refactoring), but might make this message go away (if it is still there).

Optimizations made some changes to how tables are rendered, so please test if there are any undesired side effects (I have not seen any on my side but I did not have a chance to do a very extensive testing).

The test package is https://github.com/alpheios-project/webextension/archive/infltbl-i170.zip.

**Comment by balmas** [comment]
seems better to me.  waiting for feedback from @monzug

**Comment by monzug** [comment]
just installed on Firefox. Unfortunately I got the message at first lookup for cepit from this site https://it.wikipedia.org/wiki/Quarta_coniugazione_latina
Definitely there is an improvement in loading speed.
will try more sites and words

**Comment by monzug** [comment]
I got the yellow message 3 times out of 4. speed has slightly improved.

**Comment by balmas** [comment]
ok, so @kirlat it seems like perhaps we need to try the vue.js lazy rendering.

**Comment by balmas** [comment]
this build has also exposed a bug displaying greek pronoun tables:

![screenshot from 2018-09-20 16-52-02](https://user-images.githubusercontent.com/644870/45846343-8b9afd80-bcf5-11e8-8936-6c75d6b1d2e1.png)

**Comment by kirlat** [comment]
I've noticed this bug too and submitted an updated build yesterday. Hope it will fix it. I'll do more testing to make sure everything is fine. 

If you still experience problems, please let me know what words caused it.

I will work on changing timing for inflection browser table rendering to avoid all of them being rendered at once.

**Comment by kirlat** [comment]
Please check an updated build where inflection browser tables are rendered at the moment when user clicks on a table title in order to expand its content: https://github.com/alpheios-project/webextension/archive/infltbl-i170.zip. Does it behave better this way?

**Comment by monzug** [comment]
much better!!!! still slower that what it used to be. I got the message
twice out of 6 sites that I tried. this is one site:
https://it.wikiversity.org/wiki/I_verbi_particolari_latini_(superiori)#Eo,_is,_ivi_/_ii,_ire_(Andare)

I did also try in clean environment (opening the above site in a new tab)
and different words each time. it look like cepit and excrucior are the
culprit. 'capit' worked fine, so did 'caro', 'iremus' (very slow),

**Comment by kirlat** [comment]
Glad that it's better! Let me check these particular words and see if I can optimize them more.

**Comment by monzug** [comment]
hope it can help. let me know if this can be a good approach and I will try
testing with more words.

**Comment by kirlat** [comment]
Please check the latest variant: https://github.com/alpheios-project/webextension/archive/infltbl-i170.zip.

Trying to do the best I can without reworking too much of the code to avoid extensive re-testing it will cause. Please let me know if this one shows any improvements.

**Comment by monzug** [comment]
ok. going to try right now.

**Comment by monzug** [comment]
uhm, not any better. 3 out of 4 latin sites got the script message. Greek
sites worked fine with both words and verbs lookups.

**Comment by kirlat** [comment]
Hmm, interesting. Can you please specify what site and what word it is?

**Comment by balmas** [comment]
@kirlat maybe we could try taking 5 of the 6 regular verb tables out of the browser for now?  It seems that this started when we added those in, and I would be willing to sacrifice the various views of them in the browser in order to postpone major rework right now.

**Comment by kirlat** [comment]
@balmas, we could do that, but I want to be sure that it's them to blame first. This already has been reworked, so inflection browser tables should not take much processing time, but my test cases could be different. I'm wondering if this is something else causing this. Will try to record performance one the problematic sites.

**Comment by monzug** [comment]
same as reported above with verb cepit.

**Comment by monzug** [comment]
it works fine as new tab. could it be the reload? after installing the new
build and the site is already open in the browser, I need to reload the
page fort he webextension to work.

**Comment by monzug** [comment]
Take it back. I did try few times with new tab and I got few script
messages.

**Comment by balmas** [comment]
Just to be sure there isn't lingering code contributing, I would suggest restarting your browser and testing then.

**Comment by monzug** [comment]
ok

**Comment by monzug** [comment]
will also reinstall latest build with @kirlat fix

**Comment by kirlat** [comment]
Tried the word "cepit" on https://it.wikiversity.org/wiki/I_verbi_particolari_latini_(superiori)#Eo,_is,_ivi_/_ii,_ire_(Andare) and it does not show a warning for me (tried several times and no message so far).

I used FF 62.0. I enabled add-on debugging and installed an extension from file. Extension was located on my local drive in webextensions/dist folder, it was the last build I've submitted.

**Comment by balmas** [comment]
I am wondering if this is possibly not related to the inflection tables now. It could just be that with those loading faster, we are identifying a differet performance lag with Latin lookups.  When I look at the debugger while the lexical data is loading message is displaying and hanging it's actually well past the inflection table initialization phase and is waiting on loading of definitions.  This would actually make sense with the fact that it occurs only with certain words.

**Comment by balmas** [comment]
After investigating further, I'm 99% convinced that the remaining performance problem is unrelated to the inflection tables. I believe it has to do with the wait on the load of the lemma index which only occurs upon the first lookup. I think the problem was probably there for quite some time, almost certainly in the alpha release, but the intense focus on the initial lookup for this bug has brought our attention to it.  I have made a small fix in the lexicon client code (see https://github.com/alpheios-project/lexicon-client/issues/16) which is a small improvement, but which doesn't address the problem of waiting on the index. That is something I would like to postpone until after the beta release if possible.  I believe this problem is largely limited to the very first lookup upon activation and we can list it as an outstanding issue in the release notes.

**Comment by balmas** [comment]
can be tested in https://github.com/alpheios-project/webextension/tree/qa-2.0.3-10/dist

**Comment by monzug** [comment]
I do still get the script message at first  lookup for some words that take longer to upload e.g. vinceremus. I have to say that it's not limited to first lookup but this is less common.

**Comment by balmas** [comment]
@monzug reports that she cannot reproduce the Firefox script message with the original build for this issue, and can regularly with QA build 10 so it seems there must be something we added in between that exacerbates it. Want to take this and investigate a bit.

**Comment by balmas** [comment]
Ok, fairly certain that the performance is in the inflection tables. I removed the calls to create the inflection table from the query and the controller and it gets very speedy again. I added them back in one by one:

`ViewSetFactory.hasInflectionsEnabled(languageID)` has no discernible impact on speec
`ViewSetFactory.create(homonym) introduces one slowdown and then when we actually add the inflectionsViewSet to the panel component, it gets *really* slow.

continuing investigation...

**Comment by monzug** [comment]
if it can help, version #40 did not have the regular latin verb tables.

**Comment by balmas** [comment]
yes, it's definitely the creation of the inflection tables for the browser and adding the regular verb tables probably just put it over the top. Working on a fix now.

**Comment by balmas** [comment]
Ok, please test this in https://github.com/alpheios-project/webextension/tree/qa-2.0.3-11/dist

What you should see is a slightly longer time for the panel to upon upon activation but significantly less delay on lookups.

We should now be creating the inflection browser tables only once upon activation.

**Comment by monzug** [comment]
YES!!!! didn't get the message and the speed of lookup has improved
dramatically! brava. will do more testing tomorrow

**Comment by balmas** [comment]
There is another fix for this in https://github.com/alpheios-project/webextension/tree/qa-2.0.3-12/dist

I have delayed the loading of the inflection browser until after the panel is already displaying, so that it doesn't give the impression of a slow activation. It does mean the panel icons change shortly after it first loads, but I think that's better than slowing down the initial display of it.

**Comment by monzug** [comment]
Looks good to me: no script messages and no speed issues.
